keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -333,36 +337,302 @@ or later. The dual-stack feature requires no changes to this API.
+  ipFamilies: # the order of the families is according to how they are configured on the cluster
+    - IPv4 # cluster default
+    - IPv6 
+  clusterIP: <ipv4> # allocated
+  clusterIPs:
+    - <ipv4>
+    - <ipv6> 
+   <snip>
+```
+
+Kubernetes default service is expressed current in pods as environment variables, these variables offers location, ip and other details of kubernetes default service. These variables will remain. The following variables will be added
+
+```
+# for IPv4
+KUBERNETES_IPv4_PORT=tcp://<IPv4>:443
+KUBERNETES_IPv4_PORT_443_TCP=tcp://<IPv4>:443
+KUBERNETES_IPv4_PORT_443_TCP_ADDR=<IPv4>
+KUBERNETES_IPv4_PORT_443_TCP_PORT=443
+KUBERNETES_IPv4_PORT_443_TCP_PROTO=tcp
+KUBERNETES_IPv4_SERVICE_HOST<IPv4>
+KUBERNETES_IPv4_SERVICE_PORT=443
+KUBERNETES_IPv4_SERVICE_PORT_HTTPS=443
+ 
+
+# for IPv6
+KUBERNETES_IPv64_PORT=tcp://<IPv6>:443
+KUBERNETES_IPv4_PORT_443_TCP=tcp://<IPv6>:443
+KUBERNETES_IPv4_PORT_443_TCP_ADDR=<IPv6>
+KUBERNETES_IPv4_PORT_443_TCP_PORT=443
+KUBERNETES_IPv4_PORT_443_TCP_PROTO=tcp
+KUBERNETES_IPv4_SERVICE_HOST<IPv6>
+KUBERNETES_IPv4_SERVICE_PORT=443
+KUBERNETES_IPv4_SERVICE_PORT_HTTPS=443
+``` 
 
 #### Type NodePort, LoadBalancer, ClusterIP
-Services will only be assigned an address from a single address family (either IPv4 or IPv6). The address family for the Service’s cluster IP is determined by setting the field, `.spec.ipFamily`, on that Service. You can only set this field when creating a new Service. Setting the `.spec.ipFamily` field is optional and should only be used if you plan to enable IPv4 and IPv6 Services and Ingresses on your cluster. The configuration of this field not a requirement for egress traffic. The default address family for your cluster is the address family of the first service cluster IP range configured via the `--service-cluster-ip-range` flag to the kube-controller-manager. The `.spec.ipFamily` field can be set to either IPv4 or IPv6.
+In a dualstack cluster Services can be:
+1. Single Stack: service ip are allocated (or reserved, if provided by user) from the first or the second cidr (if configured) entry as configured using `--service-cluster-ip-range` flag, in this scenario services will have a single allocated `ClusterIP` (also set in `ClusterIPs`).
+2. Dual Stack (optional): service ip is allocated (or reserved, if provided by user) from both the first and the second cidr (if configured) entries. In this scenario services will have a dual assigned `ClusterIPs`. If the cluster is not configured for dualstack then  the resulting service will be created as a single stack and `ClusterIP/ClusterIPs` will be assigned from a single family.
+3. Dual Stack (required): if the cluster is not configured for dual stack then service creation will fail. If the cluster is configured for dual stack the resulting service will carry two dual stack `ClusterIPs`.
+
+The above is achieved using the following changes to Service api.
+1. Add a `spec.preferDualStack` optional boolean field.
```

Comment:
we need to spec the mutability of these, and specifically how a user can move from a service that is (false, [x]) to (true, [x, y])

Reply:
added